### PR TITLE
feat: add new support for testing with environment variables

### PR DESCRIPTION
**Pull request by devbww**
`EnvironmentRestorer` is an RAII object that will, upon destruction,
restore the previous state of environment variables that it modified.
For example:
```
  // ${VAR1} and ${VAR2} have some initial state.
  {
    EnvironmentRestorer env;
    env.SetEnv("VAR1", "value");
    env.UnsetEnv("VAR2");
    // ${VAR1} now holds "value1" and ${VAR2} is unset.
  }
  // The initial state of ${VAR1} and ${VAR2} have been restored.
```
[Note: The idea is for this to replace `EnvironmentVariableRestore`.]

---
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp-common/180)

### CODE REVIEW

File: google/cloud/testing_util/environment_restorer.h [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+namespace cloud {
+inline namespace GOOGLE_CLOUD_CPP_NS {
+namespace testing_util {
+
+/**
+ * Helper class to restore the value of environment variables.
+ */
+class EnvironmentRestorer {
+ public:
+  EnvironmentRestorer() = default;
```
Comment (devjgm) [inline google/cloud/testing_util/environment_restorer.h]: Most of our current uses of `EnvironmentVariableRestore` are used to restore a single environment variable. I think it'd be useful to add a constructor to this class like: `EnvironmentRestorer(std::string const& variable, optional<std::string> value)` to let users set/unset a single variable in a single statement. What do you think?

File: google/cloud/testing_util/environment_restorer.h [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+namespace google {
+namespace cloud {
+inline namespace GOOGLE_CLOUD_CPP_NS {
+namespace testing_util {
+
+/**
+ * Helper class to restore the value of environment variables.
+ */
+class EnvironmentRestorer {
```
Comment (devjgm) [inline google/cloud/testing_util/environment_restorer.h]: I personally find the word "Restorer" to be a little bit of a mental tongue twister that I often typo (it's probably just me, though). Consider if you'd prefer a name like `ScopedEnvironment` or `ScopedEnv` or something. Up to you. The name you have here is fine with me if that's what you prefer.

File: google/cloud/testing_util/environment_restorer.h (hunk 1 above)
Comment (devbww) [reply google/cloud/testing_util/environment_restorer.h]: Done.

File: google/cloud/testing_util/environment_restorer.h (hunk 2 above)
Comment (devbww) [reply google/cloud/testing_util/environment_restorer.h]: Done (`ScopedEnvironment`).

**Pull request merged**